File: apps/api/app/domain/semantic_cache.py
"""Qdrant semantic cache — same PointStruct / query_points pattern as the RAG notebooks."""

from __future__ import annotations

import logging
import re
import time
import uuid
from dataclasses import dataclass
from typing import Any, Literal

from qdrant_client import QdrantClient
from qdrant_client.models import (
    Distance,
    FieldCondition,
    Filter,
    FilterSelector,
    MatchValue,
    PayloadSchemaType,
    PointStruct,
    VectorParams,
)

from app.core.config import get_settings
from app.domain.embeddings import embed_query, embedding_configured, embedding_dim, get_embedding

logger = logging.getLogger(__name__)

Mode = Literal["full", "hybrid", "miss"]


@dataclass
class SemanticEntry:
    id: str
    prompt: str
    answer: str
    model: str
    tier: str
    quality: float
    created_at: float


@dataclass
class SemanticMatch:
    entry: SemanticEntry
    similarity: float
    mode: Mode
    novel: str
    shared_ratio: float


_client: QdrantClient | None = None
_ensured = False


def semantic_enabled() -> bool:
    settings = get_settings()
    raw = (settings.semantic_cache or "true").strip().lower()
    return raw not in {"0", "false", "off"}


def qdrant_configured() -> bool:
    settings = get_settings()
    return bool(settings.qdrant_url.strip()) and embedding_configured()


def collection_name() -> str:
    return get_settings().qdrant_collection.strip() or "promptimizer_semantic_2048"


def _client_get() -> QdrantClient:
    """Notebook style: QdrantClient(url=...)."""
    global _client
    if _client is not None:
        return _client
    settings = get_settings()
    kwargs: dict[str, Any] = {
        "url": settings.qdrant_url.rstrip("/"),
        "prefer_grpc": False,
    }
    if settings.qdrant_api_key.strip():
        kwargs["api_key"] = settings.qdrant_api_key.strip()
    _client = QdrantClient(**kwargs)
    return _client


def ensure_collection() -> bool:
    """create_collection(VectorParams(size=2048, distance=Distance.COSINE)) like the notebook."""
    global _ensured
    if not qdrant_configured():
        return False
    if _ensured:
        return True

    qdrant_client = _client_get()
    name = collection_name()
    dim = embedding_dim()
    existing = {c.name for c in qdrant_client.get_collections().collections}

    if name not in existing:
        qdrant_client.create_collection(
            collection_name=name,
            vectors_config=VectorParams(size=dim, distance=Distance.COSINE),
        )
    else:
        info = qdrant_client.get_collection(name)
        params = info.config.params.vectors
        size = int(params.size) if hasattr(params, "size") else None
        if size is not None and size != dim:
            raise RuntimeError(
                f'Qdrant collection "{name}" is {size}-d but embeddings are {dim}-d. '
                "Set QDRANT_COLLECTION to a new name."
            )

    try:
        qdrant_client.create_payload_index(
            collection_name=name,
            field_name="owner",
            field_schema=PayloadSchemaType.KEYWORD,
        )
    except Exception as err:  # noqa: BLE001
        msg = str(err).lower()
        if "already" not in msg and "conflict" not in msg and "exists" not in msg:
            logger.warning("payload index owner: %s", err)

    _ensured = True
    return True


def _normalize(text: str) -> str:
    s = text.lower().strip()
    s = re.sub(r"[.…]+$", "", s)
    s = re.sub(r"\s+", " ", s)
    return s


def _canonicalize(text: str) -> str:
    s = _normalize(text)
    s = s.replace("×", "*").replace("x", "*")
    s = re.sub(r"\btimes\b", "*", s)
    s = re.sub(r"\bmultiplied by\b", "*", s)
    s = re.sub(r"[^0-9*+\-/\s]", " ", s)
    s = re.sub(r"\s+", " ", s).strip()
    nums = re.findall(r"\d+", s)
    if len(nums) >= 2 and "*" in s:
        return f"{nums[0]} * {nums[1]}"
    return _normalize(text)


def _entities_compatible(a: str, b: str) -> bool:
    if _canonicalize(a) and _canonicalize(a) == _canonicalize(b):
        return True
    na = set(re.findall(r"\d+", a))
    nb = set(re.findall(r"\d+", b))
    if na or nb:
        return na == nb
    neg_a = bool(re.search(r"\bnot\b|n't\b", a.lower()))
    neg_b = bool(re.search(r"\bnot\b|n't\b", b.lower()))
    if neg_a != neg_b:
        return False
    return True


def _threshold() -> float:
    return float(get_settings().semantic_threshold)


def _full_hit() -> float:
    return float(get_settings().semantic_full_hit)


def _paraphrase_hit() -> float:
    return float(get_settings().semantic_paraphrase_hit)


def _pick(
    prompt: str,
    candidates: list[tuple[SemanticEntry, float]],
) -> SemanticMatch | None:
    if not candidates:
        return None
    prompt_norm = _normalize(prompt)
    prompt_canon = _canonicalize(prompt)
    best_any: tuple[SemanticEntry, float, bool, bool, bool] | None = None
    best_safe: tuple[SemanticEntry, float, bool, bool, bool] | None = None

    for entry, raw_sim in candidates:
        same_norm = bool(prompt_norm and prompt_norm == _normalize(entry.prompt))
        same_canon = bool(prompt_canon and prompt_canon == _canonicalize(entry.prompt))
        compatible = same_norm or same_canon or _entities_compatible(prompt, entry.prompt)
        sim = 1.0 if same_norm or same_canon else raw_sim
        cand = (entry, sim, compatible, same_canon, same_norm)
        if best_any is None or sim > best_any[1]:
            best_any = cand
        if compatible and (best_safe is None or sim > best_safe[1]):
            best_safe = cand

    pick = best_safe or best_any
    if not pick:
        return None
    entry, sim, compatible, same_canon, same_norm = pick
    shared_ratio = 1.0 if same_norm or same_canon else 0.5
    novel = "" if same_norm or same_canon else prompt

    if not compatible:
        return SemanticMatch(entry, sim, "miss", prompt, shared_ratio)

    if same_norm or same_canon or sim >= _full_hit() or (sim >= _paraphrase_hit() and compatible):
        return SemanticMatch(
            entry,
            1.0 if same_norm or same_canon else sim,
            "full",
            "",
            shared_ratio,
        )
    if sim >= _threshold():
        return SemanticMatch(entry, sim, "hybrid", novel, shared_ratio)
    return SemanticMatch(entry, sim, "miss", prompt, shared_ratio)


def _entry_from_payload(point_id: Any, payload: dict[str, Any]) -> SemanticEntry | None:
    if not payload.get("answer") or not payload.get("prompt"):
        return None
    return SemanticEntry(
        id=str(payload.get("entry_id") or point_id),
        prompt=str(payload["prompt"]),
        answer=str(payload["answer"]),
        model=str(payload.get("model") or ""),
        tier=str(payload.get("tier") or ""),
        quality=float(payload.get("quality") or 0),
        created_at=float(payload.get("created_at") or 0),
    )


async def find_similar(prompt: str, owner: str) -> SemanticMatch | None:
    """Notebook retrieve: embed → query_points → read payload + score from results.points."""
    if not semantic_enabled() or not qdrant_configured() or not owner.strip():
        return None
    try:
        ensure_collection()
        query_embedding = await embed_query(prompt)
        qdrant_client = _client_get()
        results = qdrant_client.query_points(
            collection_name=collection_name(),
            query=query_embedding,
            limit=8,
            with_payload=True,
            query_filter=Filter(
                must=[FieldCondition(key="owner", match=MatchValue(value=owner))]
            ),
        )

        candidates: list[tuple[SemanticEntry, float]] = []
        for result in results.points or []:
            payload = result.payload or {}
            entry = _entry_from_payload(result.id, payload)
            if not entry:
                continue
            candidates.append((entry, float(result.score or 0)))
        return _pick(prompt, candidates)
    except Exception as err:  # noqa: BLE001
        logger.warning("semantic find_similar failed: %s", err)
        return None


async def remember_semantic(
    *,
    prompt: str,
    answer: str,
    model: str,
    tier: str,
    quality: float,
    owner: str,
) -> None:
    """Notebook upsert: PointStruct(id, vector, payload) → upsert(wait=True, points=[...])."""
    if not semantic_enabled() or not qdrant_configured():
        return
    if not prompt.strip() or not answer.strip() or not owner.strip():
        return
    try:
        ensure_collection()
        embedding = await embed_query(prompt)
        entry_id = f"sem_{uuid.uuid4().hex[:12]}"
        point_id = str(uuid.uuid4())

        point = PointStruct(
            id=point_id,
            vector=embedding,
            payload={
                "prompt": prompt[:12_000],
                "answer": answer[:24_000],
                "model": model,
                "tier": tier,
                "quality": quality,
                "owner": owner,
                "created_at": time.time() * 1000,
                "entry_id": entry_id,
                "embed_model": get_settings().embedding_model or "nvidia/nemotron-3-embed-1b",
            },
        )

        qdrant_client = _client_get()
        qdrant_client.upsert(
            collection_name=collection_name(),
            wait=True,
            points=[point],
        )
    except Exception as err:  # noqa: BLE001
        logger.warning("semantic remember failed: %s", err)


async def delete_semantic_by_owner(owner: str) -> None:
    if not qdrant_configured() or not owner.strip():
        return
    try:
        ensure_collection()
        qdrant_client = _client_get()
        qdrant_client.delete(
            collection_name=collection_name(),
            points_selector=FilterSelector(
                filter=Filter(
                    must=[FieldCondition(key="owner", match=MatchValue(value=owner))]
                )
            ),
        )
    except Exception as err:  # noqa: BLE001
        logger.warning("semantic delete failed: %s", err)


def build_hybrid_messages(
    messages: list[dict[str, Any]],
    match: SemanticMatch,
) -> list[dict[str, Any]]:
    system_extra = "\n".join(
        [
            "You are continuing a related request. A similar prior answer is cached below.",
            "Reuse correct shared reasoning. Focus compute on the NEW / dissimilar parts.",
            f"Similarity: {match.similarity * 100:.1f}%.",
            "",
            "=== CACHED SIMILAR PROMPT ===",
            match.entry.prompt[:4000],
            "",
            "=== CACHED ANSWER ===",
            match.entry.answer[:6000],
            "",
            "=== NOVEL / DISSIMILAR FOCUS ===",
            (match.novel[:4000] or "(minor wording change — adapt the cached answer carefully)"),
        ]
    )
    out = [dict(m) for m in messages]
    for i, m in enumerate(out):
        if m.get("role") == "system":
            out[i] = {**m, "content": f"{m.get('content', '')}\n\n{system_extra}"}
            return out
    out.insert(0, {"role": "system", "content": system_extra})
    return out


# Sync helpers matching notebook cell names (useful for scripts / smoke tests).
def retrieve_data(query: str, owner: str, k: int = 5) -> dict[str, Any]:
    """Sync retrieve like 03-RAG-pipeline.retrieve_data."""
    ensure_collection()
    query_embedding = get_embedding(query)
    results = _client_get().query_points(
        collection_name=collection_name(),
        query=query_embedding,
        limit=k,
        with_payload=True,
        query_filter=Filter(
            must=[FieldCondition(key="owner", match=MatchValue(value=owner))]
        ),
    )
    prompts: list[str] = []
    answers: list[str] = []
    scores: list[float] = []
    for result in results.points:
        payload = result.payload or {}
        prompts.append(str(payload.get("prompt") or ""))
        answers.append(str(payload.get("answer") or ""))
        scores.append(float(result.score or 0))
    return {
        "retrieved_prompts": prompts,
        "retrieved_answers": answers,
        "similarity_scores": scores,
    }

File: apps/web/src/server/semantic-cache.ts
import { cacheGet, cacheSet, userCacheKey } from "./upstash";

export type SemanticEntry = {
  id: string;
  prompt: string;
  embedding: number[];
  answer: string;
  model: string;
  tier: string;
  quality: number;
  created_at: number;
};

export type SemanticMatch = {
  entry: SemanticEntry;
  similarity: number;
  mode: "full" | "hybrid" | "miss";
  novel: string;
  shared_ratio: number;
};

const DIM = 256;

function indexKey(owner?: string | null) {
  return userCacheKey(owner, "semantic", "index");
}

function clamp01(n: number) {
  return Math.max(0, Math.min(1, n));
}

/** Env: SEMANTIC_THRESHOLD (default 0.5), SEMANTIC_FULL_HIT (default 0.92). */
export function semanticThreshold() {
  const n = Number(process.env.SEMANTIC_THRESHOLD ?? 0.5);
  return Number.isFinite(n) ? clamp01(n) : 0.5;
}

export function semanticFullHit() {
  const n = Number(process.env.SEMANTIC_FULL_HIT ?? 0.92);
  return Number.isFinite(n) ? clamp01(n) : 0.92;
}

export function semanticEnabled() {
  const raw = (process.env.SEMANTIC_CACHE ?? "true").trim().toLowerCase();
  return raw !== "0" && raw !== "false" && raw !== "off";
}

function hashToken(token: string) {
  let h = 2166136261;
  for (let i = 0; i < token.length; i += 1) {
    h ^= token.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  return h >>> 0;
}

/**
 * Canonical form for prompt-cache keys + embeddings.
 * Strips trailing ellipsis/punctuation so "What is 17 * 24?..." ≡ "What is 17 * 24?".
 */
export function normalizeCachePrompt(text: string): string {
  return text
    .normalize("NFKC")
    .replace(/\u2026/g, "...")
    .replace(/[×✕✖⨯]/g, "*")
    .toLowerCase()
    .replace(/\s+/g, " ")
    .trim()
    .replace(/(?:\.{2,}|…+|[.!?]+)+\s*$/g, "")
    .replace(/\s+/g, " ")
    .trim();
}

/** Lightweight hashed bag-of-tokens + char-trigrams → fixed vector (no external embed API). */
export function embedText(text: string): number[] {
  const vec = new Float64Array(DIM);
  const normalized = normalizeCachePrompt(text);
  if (!normalized) return Array.from(vec);

  const tokens = normalized.split(/[^a-z0-9_+./-]+/).filter(Boolean);
  for (const token of tokens) {
    const idx = hashToken(token) % DIM;
    vec[idx] += 1;
    if (token.length >= 4) {
      const idx2 = hashToken(`##${token.slice(0, 4)}`) % DIM;
      vec[idx2] += 0.5;
    }
  }
  for (let i = 0; i < normalized.length - 2; i += 1) {
    const tri = normalized.slice(i, i + 3);
    if (/\s/.test(tri)) continue;
    vec[hashToken(`tri:${tri}`) % DIM] += 0.25;
  }

  let norm = 0;
  for (let i = 0; i < DIM; i += 1) norm += vec[i] * vec[i];
  norm = Math.sqrt(norm) || 1;
  const out = new Array<number>(DIM);
  for (let i = 0; i < DIM; i += 1) out[i] = vec[i] / norm;
  return out;
}

export function cosineSimilarity(a: number[], b: number[]) {
  const n = Math.min(a.length, b.length);
  if (!n) return 0;
  let dot = 0;
  let na = 0;
  let nb = 0;
  for (let i = 0; i < n; i += 1) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  const denom = Math.sqrt(na) * Math.sqrt(nb);
  return denom ? clamp01(dot / denom) : 0;
}

/** Sentences/clauses in `prompt` that are weakly covered by `cachedPrompt`. */
export function extractNovelParts(prompt: string, cachedPrompt: string): { novel: string; shared_ratio: number } {
  const parts = prompt
    .split(/(?<=[.!?])\s+|\n+/)
    .map((p) => p.trim())
    .filter((p) => p.length >= 12);
  if (!parts.length) {
    return { novel: prompt.trim(), shared_ratio: 0 };
  }
  const cached = cachedPrompt.toLowerCase();
  const novelParts: string[] = [];
  let shared = 0;
  for (const part of parts) {
    const tokens = part
      .toLowerCase()
      .split(/[^a-z0-9]+/)
      .filter((t) => t.length > 3);
    if (!tokens.length) {
      novelParts.push(part);
      continue;
    }
    const hit = tokens.filter((t) => cached.includes(t)).length / tokens.length;
    if (hit >= 0.65) shared += 1;
    else novelParts.push(part);
  }
  const shared_ratio = shared / parts.length;
  return {
    novel: novelParts.join(" ").trim() || prompt.trim(),
    shared_ratio,
  };
}

async function loadIndex(owner?: string | null): Promise<SemanticEntry[]> {
  const rows = (await cacheGet<SemanticEntry[]>(indexKey(owner))) ?? [];
  return Array.isArray(rows) ? rows : [];
}

async function saveIndex(owner: string | null | undefined, rows: SemanticEntry[]) {
  const max = Math.max(32, Number(process.env.SEMANTIC_CACHE_SIZE ?? 200) || 200);
  await cacheSet(indexKey(owner), rows.slice(-max));
}

export async function findSimilar(prompt: string, owner?: string | null): Promise<SemanticMatch | null> {
  if (!semanticEnabled()) return null;
  const embedding = embedText(prompt);
  const promptNorm = normalizeCachePrompt(prompt);
  const index = await loadIndex(owner);
  if (!index.length) return null;

  let best: SemanticEntry | null = null;
  let bestSim = 0;
  for (const entry of index) {
    if (!entry?.answer) continue;
    // Re-embed from stored prompt so lookup stays correct after normalize/embed tweaks.
    const entryVec = entry.prompt ? embedText(entry.prompt) : entry.embedding;
    if (!entryVec?.length) continue;
    const sameNorm = promptNorm && promptNorm === normalizeCachePrompt(entry.prompt);
    const sim = sameNorm ? 1 : cosineSimilarity(embedding, entryVec);
    if (sim > bestSim) {
      bestSim = sim;
      best = entry;
    }
  }
  if (!best) return null;

  const threshold = semanticThreshold();
  const full = semanticFullHit();
  const { novel, shared_ratio } = extractNovelParts(prompt, best.prompt);
  const sameAsBest = Boolean(promptNorm && promptNorm === normalizeCachePrompt(best.prompt));

  if (bestSim >= full || sameAsBest) {
    return { entry: best, similarity: sameAsBest ? 1 : bestSim, mode: "full", novel: "", shared_ratio };
  }
  if (bestSim >= threshold) {
    return { entry: best, similarity: bestSim, mode: "hybrid", novel, shared_ratio };
  }
  return { entry: best, similarity: bestSim, mode: "miss", novel: prompt, shared_ratio };
}

export async function rememberSemantic(input: {
  prompt: string;
  answer: string;
  model: string;
  tier: string;
  quality: number;
  owner?: string | null;
}) {
  if (!semanticEnabled()) return;
  if (!input.prompt.trim() || !input.answer.trim()) return;
  const entry: SemanticEntry = {
    id: `sem_${crypto.randomUUID().replaceAll("-", "").slice(0, 12)}`,
    prompt: input.prompt.slice(0, 12_000),
    embedding: embedText(input.prompt),
    answer: input.answer.slice(0, 24_000),
    model: input.model,
    tier: input.tier,
    quality: input.quality,
    created_at: Date.now(),
  };
  const index = await loadIndex(input.owner);
  index.push(entry);
  await saveIndex(input.owner, index);
}

export function buildHybridMessages(
  textMessages: Array<{ role: string; content: string }>,
  match: SemanticMatch,
): Array<{ role: string; content: string }> {
  const systemExtra = [
    "You are continuing a related request. A similar prior answer is cached below.",
    "Reuse correct shared reasoning. Focus compute on the NEW / dissimilar parts.",
    `Similarity: ${(match.similarity * 100).toFixed(1)}%. Shared-ratio: ${(match.shared_ratio * 100).toFixed(0)}%.`,
    "",
    "=== CACHED SIMILAR PROMPT ===",
    match.entry.prompt.slice(0, 4000),
    "",
    "=== CACHED ANSWER ===",
    match.entry.answer.slice(0, 6000),
    "",
    "=== NOVEL / DISSIMILAR FOCUS ===",
    match.novel.slice(0, 4000) || "(minor wording change — adapt the cached answer carefully)",
  ].join("\n");

  const out = [...textMessages];
  const sysIdx = out.findIndex((m) => m.role === "system");
  if (sysIdx >= 0) {
    out[sysIdx] = { ...out[sysIdx], content: `${out[sysIdx].content}\n\n${systemExtra}` };
  } else {
    out.unshift({ role: "system", content: systemExtra });
  }
  return out;
}
